### DISCUSSION TITLE
Where to redirect with notFound when using generateMetadata?

### DISCUSSION THREAD

**Discussion by codinginflow** [vercel/next.js#49819] [App Router]
Some of my pages export `generateMetadata` functions that fetch the same data as the Page component itself.

If the data is not found (indicated by a 404 server response), do I need to redirect (by calling the `notFound()` function) in both functions?

If I only redirect in the Page component, it doesn't work. Instead, I'm getting to the generic error page (because the error is not caught). Everything seems to work **if I **only** catch + redirect in `generateMetadata`**.

My question: Is it okay to only catch and redirect 404 responses in `generateMetadata` or do I need to do it in both functions?

```
import * as UsersApi from "@/network/api/users";
import { notFound } from "next/navigation";
import { NotFoundError } from "@/network/http-errors";
import UserProfilePage from "./UserProfilePage";
import { Metadata } from "next";
import { cache } from "react";

interface PageProps {
    params: { username: string; }
};

const getUser = cache(UsersApi.getUserByUsername);

export async function generateMetadata({ params: { username } }: PageProps): Promise<Metadata> {
    try {
        const user = await getUser(username);
        return {
            title: `${user.username} - Flow Blog`,
        };
    } catch (error) {
        if (error instanceof NotFoundError) {
            notFound();
        } else {
            throw error;
        }
    }
}

export default async function Page({ params: { username } }: PageProps) {
    **// Do I need to catch + redirect 404s here as well?**
    const user = await getUser(username);
    return <UserProfilePage user={user} />;
}
```

**Comment by CanRau** [comment]
Just hit the same, had to call `notFound` in both, kinda makes sense, kinda confusing.
Even more confusing that I, so far, can't get the segment's `not-found.tsx` working, it's rendering `app/not-found.tsx` instead 🤔 

Update: also moved it up a level next to the dynamic segment instead of next to the segments `page.tsx` still the same, though according to the `console.log`s I've put in both `not-found.tsx` both get called, though `app/not-found.tsx` seems to win
